File: Python3/783. Minimum Distance Between BST Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        stack = deque()
        while root:
            stack.append(root)
            root = root.left

        prev = -300000
        ans = 100000
        while stack:
            node = stack.pop()
            ans = min(ans, node.val - prev)
            prev = node.val
            node = node.right
            while node:
                stack.append(node)
                node = node.left

        return ans

File: Java/783. Minimum Distance Between BST Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDiffInBST(TreeNode root) {
        Stack<TreeNode> stack = new Stack();
        while (root != null) {
            stack.push(root);
            root = root.left;
        }

        int prev = -200000, ans = 100000;
        while (!stack.isEmpty()) {
            root = stack.pop();
            ans = Math.min(ans, root.val - prev);
            prev = root.val;
            root = root.right;
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
        }

        return ans;
    }
}